rclcpp/include/rclcpp/executors/executor_entities_collection.hpp

```diff
@@ -125,6 +139,125 @@ class EntityCollection
   {
     update_entities(other, *this, on_added, on_removed);
   }
+
+  // Below are some forwarded functions to the map and vector as appropriate.
+
+  typename MapType::size_type count(const Key & key) const
+  {
+    return map_.count(key);
+  }
+
+  typename MapType::iterator begin()
+  {
+    return map_.begin();
+  }
+
+  typename MapType::const_iterator begin() const
+  {
+    return map_.begin();
+  }
+
+  typename MapType::iterator end()
+  {
+    return map_.end();
+  }
+
+  typename MapType::const_iterator end() const
+  {
+    return map_.end();
+  }
+
+  typename VectorType::iterator begin_ordered()
+  {
+    return insertion_order_.begin();
+  }
+
+  typename VectorType::const_iterator begin_ordered() const
+  {
+    return insertion_order_.begin();
+  }
+
+  typename VectorType::iterator end_ordered()
+  {
+    return insertion_order_.end();
+  }
+
+  typename VectorType::const_iterator end_ordered() const
+  {
+    return insertion_order_.end();
+  }
+
+  typename MapType::const_iterator find(const Key & key) const
+  {
+    return map_.find(key);
+  }
+
+  bool empty() const noexcept
+  {
+    return insertion_order_.empty();
+  }
+
+  typename VectorType::size_type size() const noexcept
+  {
+    return insertion_order_.size();
+  }
+
+  typename MapType::iterator erase(typename MapType::const_iterator pos)
+  {
+    // from: https://en.cppreference.com/w/cpp/container/unordered_map/erase
+    // The iterator pos must be valid and dereferenceable.
+    // Thus the end() iterator (which is valid, but is not dereferenceable)
+    // cannot be used as a value for pos.
+    //
+    // Therefore we can use pos-> here safely.
+    insertion_order_.erase(
+      std::remove_if(
+        insertion_order_.begin(),
+        insertion_order_.end(),
+        [&pos](auto value) {
+          return value.first == pos->first;
+        }));
+    return map_.erase(pos);
+  }
+
+  typename VectorType::iterator erase_ordered(typename VectorType::const_iterator pos)
+  {
+    // from: https://en.cppreference.com/w/cpp/container/vector/erase
+    // The iterator pos must be valid and dereferenceable. Thus the end
+    // () iterator (which is valid, but is not dereferenceable) cannot be used
+    // as a value for pos.
+    //
+    // Therefore we can use pos-> here safely.
+
+    assert(map_.erase(pos->first) == 1);
+    return insertion_order_.erase(pos);
+  }
+
+  void clear() noexcept
+  {
+    insertion_order_.clear();
+    return map_.clear();
+  }
+
+  /// Insert into the collection and return true if inserted, otherwise false
+  /**
+   * Insertion will fail, and return false, if attempting to insert a duplicate
+   * entity.
+   */
+  [[nodiscard]]
+  bool insert(typename MapType::value_type value)
+  {
+    if (!map_.insert(value).second) {
+      // attempting to insert a duplicate entity
+      return false;
+    }
+    insertion_order_.push_back(value);
+    return true;
+  }
+
+private:
+  MapType map_;
+  VectorType insertion_order_;
```

Comment:
While not specifically related to the issue this PR is trying to solve, it would be great if an allocator could be specified for these containers. This was previously achievable with the `AllocatorMemoryStrategy`, but the `ExecutorEntitiesCollection` is wholly internal to the `Executor` class, so it's not immediately clear how the user could specify a custom allocator through the `Executor` interface.

Reply:
Yes, that's another thing I'd also like to see which didn't make it into this refactor (along with the aforementioned need for more customization of the executors that was lost with the removal of the memory strategy).

Reply:
Most likely we would not use an AllocatorT template argument, but instead support the pmr allocator pattern instead (https://en.cppreference.com/w/cpp/memory/polymorphic_allocator), based on our past experience with the allocators.